query/expression_test.go

```diff
@@ -127,7 +142,157 @@ func Test_evaluateBinaryOperation(t *testing.T) {
 			},
 			func(left, right float64) float64 { return left - right },
 			true,
-			[]float64{-3, -3, 2},
+			[][]float64{{-3, -3, 2}},
+		},
+		{
+			emptyContext,
+			"add",
+			[]Expression{
+				&LiteralSeriesExpression{
+					[]api.Timeseries{
+						api.Timeseries{
+							[]float64{1, 2, 3},
+							api.TagSet{
+								"env":  "production",
+								"host": "#1",
+							},
+						},
+						api.Timeseries{
+							[]float64{7, 7, 7},
+							api.TagSet{
+								"env":  "staging",
+								"host": "#2",
+							},
+						},
+						api.Timeseries{
+							[]float64{1, 0, 2},
+							api.TagSet{
+								"env":  "staging",
+								"host": "#3",
+							},
+						},
+					},
+				},
+				&LiteralSeriesExpression{
+					[]api.Timeseries{
+						api.Timeseries{
+							[]float64{5, 5, 5},
+							api.TagSet{
+								"env": "staging",
+							},
+						},
+						api.Timeseries{
+							[]float64{10, 100, 1000},
+							api.TagSet{
+								"env": "production",
+							},
+						},
+					},
+				},
+			},
+			func(left, right float64) float64 { return left + right },
+			true,
+			[][]float64{{11, 102, 1003}, {12, 12, 12}, {6, 5, 7}},
+		},
+		{
+			emptyContext,
+			"add",
+			[]Expression{
+				&LiteralSeriesExpression{
+					[]api.Timeseries{
+						api.Timeseries{
+							[]float64{1, 2, 3},
+							api.TagSet{
+								"env":  "production",
+								"host": "#1",
+							},
+						},
+						api.Timeseries{
+							[]float64{4, 5, 6},
+							api.TagSet{
+								"env":  "staging",
+								"host": "#2",
+							},
+						},
+						api.Timeseries{
+							[]float64{7, 8, 9},
+							api.TagSet{
+								"env":  "staging",
+								"host": "#3",
+							},
+						},
+					},
+				},
+				&LiteralSeriesExpression{
+					[]api.Timeseries{
+						api.Timeseries{
+							[]float64{2, 2, 2},
+							api.TagSet{
+								"env": "staging",
+							},
+						},
+						api.Timeseries{
+							[]float64{3, 3, 3},
+							api.TagSet{
+								"env": "staging",
+							},
+						},
+					},
+				},
+			},
+			func(left, right float64) float64 { return left * right },
+			true,
+			[][]float64{{8, 10, 12}, {14, 16, 18}, {12, 15, 18}, {21, 24, 27}},
```

Comment:
The tags do persist (see line 137 in query/expression.go), this test just does not check for them. It could be a good idea to add that information to the test.